crates/bevy_math/src/catmull_rom.rs

```diff
@@ -0,0 +1,149 @@
+        let [_, b, c, d] = segment.coeff;
+        b + c * 2.0 * t + d * 3.0 * t.powi(2)
+    }
+
+    /// Compute the instantaneous acceleration vector at `t` along the spline.
+    ///
+    /// Note that `t` varies from `0..=(n_points - 3)`.
+    pub fn acceleration(&self, t: f32) -> P {
+        let (segment, t) = self.segment(t);
+        let [_, _, c, d] = segment.coeff;
+        c * 2.0 + d * 6.0 * t
+    }
+
+    /// Returns the [`Segment`] and local `t` value given a spline's global `t` value.
+    fn segment(&self, t: f32) -> (&Segment<P>, f32) {
+        let i = (t.floor() as usize).clamp(0, self.segments.len() - 1);
+        (&self.segments[i], t - i as f32)
+    }
+
+    fn compute_segment_coefficients(
+        points: &[P],
+        tension: f32,
+    ) -> Result<Vec<Segment<P>>, CatmullRomError> {
+        if points.len() < 4 {
+            return Err(CatmullRomError::NotEnoughPoints);
+        } else {
+            Ok(points
+                .windows(4)
+                .map(|p| Self::catmull_rom_coeff([p[0], p[1], p[2], p[3]], tension))
+                .collect())
+        }
+    }
+
+    fn catmull_rom_coeff(p: [P; 4], tau: f32) -> Segment<P> {
+        Segment {
+            coeff: [
+                p[1],
+                (p[0] * -tau + p[2] * tau),
+                (p[0] * 2.0 * tau + p[1] * (tau - 3.0) + p[2] * (3.0 - 2.0 * tau) + p[3] * -tau),
+                (p[0] * -tau + p[1] * (2.0 - tau) + p[2] * (tau - 2.0) + p[3] * tau),
+            ],
+        }
+    }
+
+    /// Split the Catmull-Rom spline into `subdivisions` evenly spaced `t` values across the length
+    /// of the curve from t = `0..=1`, and sample with the supplied `sample_function`.
+    #[inline]
+    pub fn sample(&self, subdivisions: i32, sample_function: fn(&Self, f32) -> P) -> Vec<P> {
+        (0..=subdivisions)
+            .map(|i| {
+                let t = (i as f32 / subdivisions as f32) * self.segments.len() as f32;
+                sample_function(self, t)
+            })
+            .collect()
+    }
+
+    /// Split the Catmull-Rom spline into `subdivisions` evenly spaced `t` values across the length
+    /// of the curve. sampling the position at each step.
+    pub fn to_positions(&self, subdivisions: i32) -> Vec<P> {
+        self.sample(subdivisions, Self::position)
+    }
+
+    /// Split the Catmull-Rom spline into `subdivisions` evenly spaced `t` values across the length
+    /// of the curve. sampling the velocity at each step.
+    pub fn to_velocities(&self, subdivisions: i32) -> Vec<P> {
+        self.sample(subdivisions, Self::velocity)
+    }
+
+    /// Split the Catmull-Rom spline into `subdivisions` evenly spaced `t` values across the length
+    /// of the curve. sampling the acceleration at each step.
+    pub fn to_accelerations(&self, subdivisions: i32) -> Vec<P> {
+        self.sample(subdivisions, Self::acceleration)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum CatmullRomError {
```

Comment:
This should have the `Error` trait, implemented via the `thiserror` crate.

Reply:
I wasn't sure if I should pull in the dep on `bevy_math` usually we're pretty sensitive to pulling in deps. I agree though!

Reply:
Do we need sign off from `@cart` to add deps? 😆 
`bevy_math` is a super lightweight crate.

Reply:
Hmm, good question. It doesn't add it to the overall Bevy tree, so I'm much less concerned. I also think that, in general, we will continue to want to add more `Error` types to `bevy_math`.

Ping @cart?

Reply:
In the latest prototype, I've removed the need for fallible constructors, but it would still be useful to have an answer to the question.